src/main/java/org/radarbase/appserver/service/FcmNotificationService.java

```diff
@@ -1,22 +1,311 @@
+    public boolean checkIfNotificationExists(FcmNotificationDto notificationDto, String subjectId) {
+        Optional<User> user = this.userRepository.findBySubjectId(subjectId);
+        if(user.isEmpty()) {
+            throw new NotFoundException("The supplied subject ID is invalid. No user found. Please Create a User First.");
+        }
+        Notification notification = ConverterFactory.getNotificationConverter().dtoToEntity(notificationDto).setUser(user.get());
+
+        List<Notification> notifications = this.notificationRepository.findByUserId(user.get().getId());
+        return notifications.contains(notification);
+    }
+
+    //TODO : WIP
+    @Transactional(readOnly = true)
+    public FcmNotifications getFilteredNotifications(String type, boolean delivered,
+                                                     int ttlSeconds, LocalDateTime startTime,
+                                                     LocalDateTime endTime, int limit) {
         return null;
     }
-}
-    git push -u origin mastergit add
\ No newline at end of file
+
+    // TODO Also update users lastOpened metric
+    @Transactional
+    public FcmNotificationDto addNotification(FcmNotificationDto notificationDto, String subjectId, String projectId) {
+        Optional<Project> project = this.projectRepository.findByProjectId(projectId);
+        if(project.isEmpty()) {
+            throw new NotFoundException("Project Id does not exist. Please create a project with the ID first");
+        }
+
+        Optional<User> user = this.userRepository.findBySubjectIdAndProjectId(subjectId, project.get().getId());
+        if(user.isEmpty()) {
+            throw new NotFoundException("The supplied subject ID is invalid. No user found. Please Create a User First.");
+        }
+        if(!notificationRepository.existsByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
+                user.get().getId(),
+                notificationDto.getSourceId(),
+                notificationDto.getScheduledTime().toInstant(ZoneOffset.UTC),
+                notificationDto.getTitle(),
+                notificationDto.getBody(),
+                notificationDto.getType(),
+                notificationDto.getTtlSeconds()
+        )) {
+            Notification notification = ConverterFactory.getNotificationConverter().dtoToEntity(notificationDto).setUser(user.get());
+
+            notification = this.notificationRepository.save(notification);
+            this.schedulerService.scheduleNotification(notification);
+
+            return ConverterFactory.getNotificationConverter().entityToDto(notification);
+
+        } else {
+            throw new AlreadyExistsException("The Notification Already exists. Please Use update endpoint", notificationDto);
+        }
+    }
+
+    @SneakyThrows
+    @Transactional
+    public FcmNotificationDto addNotificationForced(FcmNotificationDto notificationDto, String fcmToken, String subjectId, String projectId) {
```

Comment:
What does this forcedNotification mean? When does it occur? Does it makes sense to first check whether there is notification which matches the criteria before force creating the user and project?

Reply:
This happens when messages come through the XMPP upstream. Because they don't have a REST API to first create a project and a subject so this needs to be done for that. This is never done in the REST API.

If a notification already exists then there is bound to be a project and a subject. A notification cannot exist without them.

Reply:
I am removing this in a separate PR and use the same workflow as REST for the XMPP too.